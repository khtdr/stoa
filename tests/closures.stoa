#!/usr/bin/env stoa

var counter = "outer-scope";
print counter;

var inc;
{
  var counter = 0;
  print counter;
  fun incr() { counter = counter + 1; return counter; }
  inc = incr;
}
print inc();
print inc();
print inc();

print counter;

fun makeDec(i) {
  fun dec() { i =i-1; return i; }
  return dec;
}

var i = 12;
var dec = makeDec(100);
print dec();
print dec();
print i;
