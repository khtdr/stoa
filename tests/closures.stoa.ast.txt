(program 
   (var counter outer-scope)
   (print counter)
   (var inc)
   (block 
      (var counter 0)
      (print counter)
      (fun incr (let [] (block 
         (= counter (+ counter 1))
         (return counter)
      )))
      (= inc incr)
   )
   (print (inc))
   (print (inc))
   (print (inc))
   (print counter)
   (fun makeDec (let [i] (block 
      (fun dec (let [] (block 
         (= i (- i 1))
         (return i)
      )))
      (return dec)
   )))
   (var i 12)
   (var dec (makeDec 100))
   (print (dec))
   (print (dec))
   (print i)
   (fun makeCounter (let [start] (block 
      (var n start)
      (return (let [] (block 
         (= n (+ n 1))
         (return n)
      )))
   )))
   (var c (makeCounter 212))
   (print (c))
)
