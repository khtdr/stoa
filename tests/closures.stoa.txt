outer-scope
0
1
2
3
outer-scope
99
98
12
