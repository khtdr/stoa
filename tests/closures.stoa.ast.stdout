(program 
   (var counter outer-scope)
   (print counter)
   (var inc)
   (block 
      (var counter 0)
      (print counter)
      (fun incr (let [] (block 
         (= counter (+ counter 1))
         (return counter)
      )))
      (= inc incr)
   )
   (print (call inc))
   (print (call inc))
   (print (call inc))
   (print counter)
   (fun makeDec (let [i] (block 
      (fun dec (let [] (block 
         (= i (- i 1))
         (return i)
      )))
      (return dec)
   )))
   (var i 12)
   (var dec (call makeDec 100))
   (print (call dec))
   (print (call dec))
   (print i)
   (fun makeCounter (let [start] (block 
      (var n start)
      (return (let [] (block 
         (= n (+ n 1))
         (return n)
      )))
   )))
   (var c (call makeCounter 212))
   (print (call c))
)
